### PR TITLE
ISO Header Struct Fixes

### PR DESCRIPTION
**Pull request by Thalley**
Fixes use of wrong ISO header structure, and adds a bitmask to ensure that iso header length value ignores RFU bits: 

image

### CODE REVIEW

File: subsys/usb/class/bluetooth.c [hunk 1]
```diff
@@ -195,11 +195,11 @@ static uint16_t hci_pkt_get_len(struct net_buf *buf,
...
 	}
 	case BT_BUF_ISO_OUT: {
-		struct bt_hci_iso_data_hdr *iso_hdr;
+		struct bt_hci_iso_hdr *iso_hdr;
 
 		hdr_len = sizeof(*iso_hdr);
-		iso_hdr = (struct bt_hci_iso_data_hdr *)data;
-		len = sys_le16_to_cpu(iso_hdr->slen) + hdr_len;
+		iso_hdr = (struct bt_hci_iso_hdr *)data;
+		len = sys_le16_to_cpu(bt_iso_hdr_len(iso_hdr->len)) + hdr_len;
```
Comment (koffes) [inline subsys/usb/class/bluetooth.c:202]: len + len?

File: include/bluetooth/hci.h [hunk 2]
```diff
@@ -96,8 +97,8 @@ struct bt_hci_iso_ts_data_hdr {
 #define BT_HCI_ISO_TS_DATA_HDR_SIZE     8
 
 struct bt_hci_iso_hdr {
-	uint16_t handle;
-	uint16_t len;
+	uint16_t handle; /* 12 bit handle, 2 bit PB flags, 1 bit TS_Flag */
```
Comment (koffes) [inline include/bluetooth/hci.h:100]: Should there be a comment about the RFU bit?

**Review by koffes — APPROVED** [review]
LGTM

File: subsys/usb/class/bluetooth.c (hunk 1 above)
Comment (Thalley) [reply subsys/usb/class/bluetooth.c:202]: Technically `hdr_len` should probably have been `hdr_size`, but these variable are used for ACL and CMDs as well, so I don't see it fit to change the naming of those in this PR.

File: include/bluetooth/hci.h (hunk 2 above)
Comment (Thalley) [reply include/bluetooth/hci.h]: Added

**Review by asbjornsabo — APPROVED** [review]
LGTM

File: drivers/bluetooth/hci/h4.c [hunk 3]
```diff
@@ -135,7 +135,7 @@ static inline void get_iso_hdr(void)
 	if (!rx.remaining) {
 		struct bt_hci_iso_hdr *hdr = &rx.iso;
 
-		rx.remaining = sys_le16_to_cpu(hdr->len);
+		rx.remaining = sys_le16_to_cpu(bt_iso_hdr_len(hdr->len));
```
Comment (cvinayak) [inline drivers/bluetooth/hci/h4.c:138]: It is important to have the 16-bits in machine endian-ness before applying the 14-bit mask.

```suggestion
		rx.remaining = bt_iso_hdr_len(sys_le16_to_cpu(hdr->len));
```

**Review by cvinayak — CHANGES_REQUESTED** [review]

File: drivers/bluetooth/hci/h4.c (hunk 3 above)
Comment (Thalley) [reply drivers/bluetooth/hci/h4.c:138]: You are absolutely correct - Somehow my mind went to "we must apply the Mask before converting to CPU endianess as we are applying a LE mask" but of course the endianess of the macro is always CPU endian and not LE endian... Thanks!

File: samples/bluetooth/hci_uart/src/main.c [hunk 4]
```diff
@@ -81,7 +81,8 @@ static uint32_t get_len(const uint8_t *hdr_buf, uint8_t type)
 	case H4_CMD:
 		return ((const struct bt_hci_cmd_hdr *)hdr_buf)->param_len;
 	case H4_ISO:
-		return sys_le16_to_cpu(((const struct bt_hci_iso_hdr *)hdr_buf)->len);
+		return sys_le16_to_cpu(
+			bt_iso_hdr_len(((const struct bt_hci_iso_hdr *)hdr_buf)->len));
```
Comment (cvinayak) [inline samples/bluetooth/hci_uart/src/main.c:85]: What about endian-ness convertion before header len decode?

**Review by cvinayak — CHANGES_REQUESTED** [review]

File: samples/bluetooth/hci_uart/src/main.c (hunk 4 above)
Comment (Thalley) [reply samples/bluetooth/hci_uart/src/main.c]: Forgot one - Thanks!

**Review by cvinayak — APPROVED** [review]

**Comment by Thalley** [comment]
@carlescufi Can we merge this today please? It's blocking some of our internal development.

**Review by jhedberg — APPROVED** [review]

**Comment by ycsin** [comment]
Is there anything here that should be backported?

**Comment by Thalley** [comment]
> Is there anything here that should be backported?

Not sure. ISO is still considered an experimental feature, so not sure if that really warranted. It also isn't a fatal bug, and will only be an issue with non-compliant controllers.

**Pull request merged**